### PR TITLE
Filter QL lex, parser, dialect

### PR DESCRIPTION
**Pull request by araddon**
Filter query language, still needs readme, etc.

### CODE REVIEW

File: expr/parse_filterql_test.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+func TestFilterQLWithJson(t *testing.T) {
+
+	ql := `
+		FILTER 
+		  AND (
+		     NAME != NULL
+		  ) 
+		WITH {
```
Comment (epsniff) [inline expr/parse_filterql_test.go]: what the WITH does is the must confusing part of the new QL.    But I suppose the Readme will cover that..

File: expr/parse_filterql.go [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+	Alias  string       // Non-Standard sql, alias/name of sql another way of expression Prepared Statement
+	With   u.JsonHelper // Non-Standard SQL for properties/config info, similar to Cassandra with, purse json
+}
+
+func NewFilterQL() *FilterQL {
+	req := &FilterQL{}
+	return req
+}
+
+type Filters struct {
```
Comment (schmichael) [inline expr/parse_filterql.go]: FilterPhrase? I dunno, `Filters` plural is a bit weird, but I see how it makes sense. Nice and succinct too. Probably just ignore me.

File: expr/parse_filterql.go [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+func (m *FilterQLTokenPager) IsEnd() bool {
+	return m.lex.IsEnd()
+}
+func (m *FilterQLTokenPager) ClauseEnd() bool {
+	tok := m.Cur()
+	//u.Debugf("IsEnd()? tok:  %v", tok)
+	switch tok.T {
+	case lex.TokenEOF, lex.TokenEOS, lex.TokenFrom, lex.TokenHaving, lex.TokenComma,
+		lex.TokenIf, lex.TokenAs, lex.TokenLimit, lex.TokenSelect:
+		return true
```
Comment (schmichael) [inline expr/parse_filterql.go]: Hm, what does this method do? Weird to see Having, If, Limit, Select, etc mentioned in filterql.

File: expr/parse_filterql.go [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+		default:
+			// first non-comment token
+			return comment
+		}
+		m.Next()
+	}
+	return comment
+}
+
+// First keyword was SELECT, so use the SELECT parser rule-set
```
Comment (schmichael) [inline expr/parse_filterql.go]: FilterQL supports Select? I don't see any tests for that...

File: expr/parse_filterql.go [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+	// one top level filter which may be nested
+	filter, err := m.parseFilters()
+	if err != nil {
+		u.Debug(err)
+		return nil, err
+	}
+	req.Filter = filter
+
+	// LIMIT
+	if err := m.parseLimit(req); err != nil {
```
Comment (schmichael) [inline expr/parse_filterql.go]: Are there tests for LIMIT clauses? I don't see any examples.

File: expr/parse_filterql_test.go (hunk 1 above)
Comment (schmichael) [inline expr/parse_filterql_test.go]: I'm used to seeing common table expressions (WITH clauses) at the beginning of queries... although this isn't a full SELECT query or traditional WITH clause so I'm not too worried about it either way.

https://msdn.microsoft.com/en-us/library/ms175972.aspx
http://www.postgresql.org/docs/8.4/static/queries-with.html

File: expr/parse_filterql.go [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+	return nil, fmt.Errorf("Did not complete parsing input: %v", m.LexTokenPager.Cur().V)
+}
+
+func (m *FilterQLParser) parseFilters() (*Filters, error) {
+
+	switch m.Cur().T {
+	case lex.TokenLogicAnd, lex.TokenAnd, lex.TokenOr, lex.TokenLogicOr:
+		// fine
+	default:
+		return nil, fmt.Errorf("Expected ( AND | OR ) but got %V", m.Cur())
```
Comment (schmichael) [inline expr/parse_filterql.go]: Is the uppercase V intentional?

File: expr/parse_filterql_test.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+	assert.Tf(t, f1.Expr != nil, "")
+	assert.Tf(t, f1.Expr.String() == "NAME != NULL", "%v", f1.Expr)
+
+	ql = `
+    FILTER
+      AND (
+          -- Lets make sure the date is good
+          daysago(datefield) < 100
+          -- as well as domain
+          , domain(url) == "google.com"
```
Comment (schmichael) [inline expr/parse_filterql_test.go]: Any chance we could easily support trailing commas? Preceding commas drive me nuts for no particular reason :)

File: expr/parse_filterql_test.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
+package expr
```
Comment (schmichael) [inline expr/parse_filterql_test.go]: Maybe add some tests against bad filterql as well to make sure it fails properly?

File: expr/parse_filterql.go (hunk 3 above)
Comment (araddon) [inline expr/parse_filterql.go]: simplified it, it is a "tokenpager" that is for recognizing ends of clauses per dialect, generally this "keyword" is part of the lexer itself but our lexer is neutral.

File: expr/parse_filterql.go (hunk 4 above)
Comment (araddon) [inline expr/parse_filterql.go]: fixed, legacy copy-pasta

File: expr/parse_filterql.go (hunk 2 above)
Comment (araddon) [inline expr/parse_filterql.go]: i renamed the `FilterQL` to `FilterStatement` which doesn't changes filters, but does help overall

File: expr/parse_filterql.go (hunk 5 above)
Comment (araddon) [inline expr/parse_filterql.go]: added some tests

File: expr/parse_filterql.go (hunk 6 above)
Comment (araddon) [inline expr/parse_filterql.go]: nope, good catch

File: expr/parse_filterql_test.go (hunk 1 above)
Comment (araddon) [inline expr/parse_filterql_test.go]: hm, maybe i should remove for now.

**Pull request merged**